MF-168 - Refactor CoAP adapter

dborovcanin opened the pull request:
1) Use NATS publish-subscribe to support CoAP observe messages
2) Use logger in sync with other services
3) Add metrics

dborovcanin committed:
Refactor CoAP adapter

1) Use NATS publish-subscribe to support CoAP observe messages
2) Use logger in sync with other services
3) Add metrics

WIP verison.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix Prometheus namespace

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add JWT authorization

WIP version.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

drasko commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+	}
+
+	// Uri-Query option ID is 15 (0xf).
+	token, err := ca.token(m.Option(0xf))
```
What's this token here? Is it CoAP token or some auth bearer token?

Because CoAP `token` is just a request ID: https://tools.ietf.org/html/rfc7252#section-5.3.1

dborovcanin replied:
Its JWT token. JWT token is sent as url-query param, which is embeded in CoAP message options, with reserved code 15. Sorry for messy code, it's WIP version.

drasko commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+
+	cid := prefix + mux.Var(m, channel)
+
+	publisher, err := ca.mc.CanAccess(cid, token)
```
Looks like CoAP token is treated like JWT

dborovcanin replied:
Since we use JWT-based auth, I found passing JWT as url query-param is pretty good workaround. Since token is wrapped as message option, it's not completely exposed. Also, obtaining token from message is straightforward and message payload stays clean. Possible problem is CoAP limitation of url query param length. 

Note: This implementation of CoAP adapter lacks self-notify checking. It will be implemented ASAP.

drasko replied:
Yes, `Uri-Query` option field looks like the good candidate for JWT passing. Let's start with this and we'll see later.

Also, maybe we should note in the comment the max size of JWT allowed (alongside with some clarifications - that we use `Uri-Query`, etc).

dborovcanin replied:
Agree, I will add some clarifications as I refactor code.

drasko reviewed (commented):
Clarification of CoAP `token` vs authorization token is needed. They represent two different concepts.

Actually - it looks like CoAP depend on DTLS for auth: https://tools.ietf.org/html/rfc7252#section-9

While we were capable to "pass" JWT to MQTT via `password` field, I am not sure if CoAP has an appropriate header in which we can put JWT. Is this solved with "Option" feature: https://tools.ietf.org/html/rfc7252#section-3.1 - I see that you are using `Uri-Query` described here: https://tools.ietf.org/html/rfc7252#section-5.10

dborovcanin committed:
Remove Zap logger dependency

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+
+// AdapterService struct represents CoAP adapter service.
+type AdapterService struct {
 	obsMap map[string][]Observer
```
Rename to `observers`. The map suffix is unnecessary since it is clear from the type that it is a map.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
-func (ca *CoAPAdapter) BridgeHandler(nm *broker.Msg) {
-	log.Printf("Received a message: %s\n", string(nm.Data))
+func (ca *AdapterService) notify(nm *broker.Msg) {
+	// TODO Authentication has to be added in order to prevent self-notifying.
```
It's not the authentication, but the origin check (if I am the publisher, don't return me my message).

dborovcanin replied:
@mijicd @drasko What do you suggest to do with notificataion messages? In this implementation, it's left to NATS, but, either way, there has to be some kind of inline structure to keep connections to be able to notify clients later. Previously, same thing was done with no NATS included at all. To keep it this way or to revert to drasko's implementation?

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+}
 
-	// And write it into the database
+func (ca *AdapterService) token(opt interface{}) (string, error) {
```
Is it possible to use the concrete type instead of `interface{}`?

dborovcanin replied:
Since [go-coap library](https://github.com/dustin/go-coap/blob/master/message.go#L366) returns an interface, eventually it's possible to do some preprocessing/checking/parsing, but it looks like there is no other easy way to get what I want from options.

mijicd replied:
Then it might be a good idea to extract the proper value before invoking this function, or if you want to keep the interface as is, don't forget to do the proper type checks.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+	if opt == nil {
+		return "", errUnauthorizedAccess
+	}
+	val := opt.(string)
```
See the comment above. This typecase surely needs both the tests and the error check.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+	return res
+}
+
+func (ca *AdapterService) registerSub(o *Observer, cid string) error {
```
Rename to `subscribe`.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+	return nil
+}
+
+func (ca *AdapterService) deregisterSub(m *coap.Message, cid string) error {
```
Rename to `unsubscribe`.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,257 @@
+
+	cid := prefix + mux.Var(m, channel)
+
+	_, err = ca.mc.CanAccess(cid, token)
```
You can inline this: 
```golang
if _, err := ca.mc.CanAccess(cid, token); err != nil {
        res.Code = coap.Unauthorized
        return res
}
```

mijicd reviewed (changes requested):
This is a first, hawk-eye review. I will make another, more thorough review as you progress with the implementation. It would be good if @drasko could provide his opinion as well.

dborovcanin committed:
Reorganize code and fix CoAP message response code

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

drasko commented on cmd/coap/main.go:
```diff
@@ -43,7 +81,8 @@ func main() {
 	go func() {
 		coapAddr := fmt.Sprintf(":%d", cfg.Port)
-		errs <- ca.Serve(coapAddr)
+		logger.Log("info", fmt.Sprintf("Start CoAP server at %s", coapAddr))
```
Align the messages to other adapters, ex. https://github.com/mainflux/mainflux/blob/master/cmd/http/main.go#L78

drasko commented on coap/utils.go:
```diff
@@ -0,0 +1,46 @@
+func convertMsg(nm *broker.Msg) (*mainflux.RawMessage, error) {
+	m := mainflux.RawMessage{}
+	if len(nm.Data) > 0 {
+		if err := json.Unmarshal(nm.Data, &m); err != nil {
```
We are using protobuf now

drasko commented on coap/api/transport.go:
```diff
@@ -0,0 +1,29 @@
+	"net"
+
+	mux "github.com/dereulenspiegel/coap-mux"
+	gocoap "github.com/dustin/go-coap"
```
`coap`, not `gocoap`, was used before - please stay consistent (change one or another)

dborovcanin replied:
One `coap` is ours, one is library package. For that reason, I can't use `coap` name. To stay consistent, I will use `gocoap` name in other packages too.

drasko reviewed (changes requested):
Please align README.md to other adapters

dborovcanin committed:
Add subscription handling using NATS

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix log messages

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

drasko commented on coap/adapter.go:
```diff
@@ -1,61 +1,129 @@
-	logger *zap.Logger
+const (
+	token    string = "token"
+	prefix   string = "src.coap."
```
Please align to latest changes - NATS topics are now equivalent to Mainflux channels: https://github.com/mainflux/mainflux/pull/207#pullrequestreview-106327364

dborovcanin committed:
Refactor CoAP adapter

Change channel prefix to correspond to latest design changes described in #204.
Add checking messages recieved from NATS.
Add pub-sub interface to  access to NATS pub-sub model.
Remove redundant dependencies from CoAP adapter and pass them as parameter to handler methods.
Improve code testability and fix log mistypes.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add CoAP adapter tests

Refactor code in order to improve testability.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix Observe method of CoAP adapter service

Remove checking payload of confirmable message.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
-const protocol string = "coap"
+// PubSub interface is used to access to NATS pub-sub model.
+type PubSub interface {
+	Publish(msg mainflux.RawMessage) error
```
No need for naming parameters. Moreover, this will have to be updated once #188 is merged.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
+
+var (
+	errBadRequest = errors.New("bad request")
+	logger        log.Logger
```
Why do you need logger variable? Let it be bound to the structure.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
-}
+// Recieve method processes message and pushes it to NATS.
+func (ca *AdapterService) Recieve(auth AuthProvider) MsgHandler {
```
Fix typo, it should be `Receive` not `Recieve`.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
+				return res
+			}
+			res.AddOption(gocoap.Observe, 0)
+		} else {
```
Either handle the condition you mentioned in the comment, or drop the entire branch. No "empty branches" should be present.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
+
+// Observe method deals with CoAP observe messages.
+func (ca *AdapterService) Observe(auth AuthProvider) MsgHandler {
+	return func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message {
```
While go enforces short names, it does not enforce meaningless names. Don't be cheap on characters :)

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,145 @@
+}
+
+// MsgHandler handles messages CoAP server recieved.
+type MsgHandler func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message
```
No need for naming the parameters here.

mijicd commented on coap/mock/nats/nats.go:
```diff
@@ -0,0 +1,81 @@
+package nats
```
Any reason for creating subpackage in `mocks`?

mijicd commented on coap/mock/manager/client.go:
```diff
@@ -0,0 +1,27 @@
+package mock
```
Stay consistent, rename this package to `mocks`.

mijicd commented on coap/mock/manager/client.go:
```diff
@@ -0,0 +1,27 @@
+
+// New method creates new mocked manager service.
+func New(email string) Client {
+	return Client{
```
Please change to `return Client{email}`.

mijicd commented on coap/mock/nats/nats.go:
```diff
@@ -0,0 +1,81 @@
+	broker "github.com/nats-io/go-nats"
+)
+
+const (
```
Please change to `const prefix = "channel."`.

mijicd commented on coap/mock/nats/nats.go:
```diff
@@ -0,0 +1,81 @@
+	prefix = "channel."
+)
+
+var log *testing.T
```
Why do you need this?

mijicd commented on coap/utils.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// ObsHandle handles observe messages and keeps connection to client in order to send notifications.
+func obsHandle(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message, offset time.Duration) broker.MsgHandler {
```
See the remark regarding naming.

mijicd commented on coap/utils.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+// AuthProvider represents ManagerClient which provides auth features.
+type AuthProvider interface {
```
Please note that some of the functions below can be bound to the `AuthProvider` instead of passing the provider to them.

mijicd reviewed (changes requested):
@drasko please take a look at my comments, and add remarks regarding the CoAP details.

dborovcanin committed:
Refactor CoAP adapter

Fix mistypes and remove parameter names in interface methods.
Attach onSubscribe method to AdapterService.
Rename mocks package and remove subpackages from it.
Pass Receive and Observe functions directly as FuncHandler.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Raise test coverage

Add tests for notification messages.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add package doc

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add Pubsub adapter and move transport handling to api package

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add channel-way subscription handling

Early WIP version

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Implement Pubsub logic using go channels

Temporary remove all unnecessary files and commented code.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Temporary remove unused packages and tests

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add server support to parse response from notification

Move CoAP server to service.go. Implementation is basically rewritten from dustin's go-coap
library, but with support to receive client's reponse to notification.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add new interfaces to support unsubscription from NATS

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Split Service interface to NATS service and adapter service

Remove Channel struct and use only chan of RawMessage to receive messages.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add Unsubscribe method to adapter service

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove obeserver from map when unsubscribe

Remove serve function from service inteface.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Move Authorize method to server

Add notify method and ticker to send confirmable messages every 24 hours.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add check if notification is sent successfully

If fail on notifying client, server will consider client is not more interested in observing resources
and therefor will remove client from observers.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add unsubscribe in case of failure sending message to clients

Add Unauthorized error code in case of unauthorized access.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix authKey method not to return status Code

Message code is generated depending on which error happened.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

anovakovic01 commented on Gopkg.lock:
```diff
@@ -392,6 +367,10 @@
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
+<<<<<<< HEAD
```
This should be resolved.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 1 {
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			err := svc.Unsubscribe(id)
+			if err != nil {
```
Can be part of `if` statement.

anovakovic01 commented on coap/adapter.go:
```diff
@@ -1,61 +1,85 @@
+		return nil
+	}
+	err := obs.Sub.Unsubscribe()
+	if err != nil {
```
This can be part of `if` statement.

drasko replied:
And use `err` here, not `ok`, in order to stay consistent.

dborovcanin replied:
Using `err` indicates an error, using `ok` indicates bool value.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+
+// MakeHandler function return new CoAP server with GET, POST and NOT_FOUND handlers.
+func MakeHandler(svc coap.Service) gocoap.Handler {
+	// auth = mgr
```
Remove commented code.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+
+func notify(svc coap.Service, id string, conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, counter *uint32, count *[]byte) error {
+	err := sendMessage(conn, addr, msg, counter, count)
+	if err != nil {
```
Can be part of `if` statement.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+			continue
+		}
+		return nil
+	}
```
Can you use circuit breaker here? Is this necessary?

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+				}
+			}
+		}
+	}()
```
Why do you need anonymous function here? Maybe you are missing `go` in front?

dborovcanin replied:
I call it as goroutine elsewhere in the code. I've used an anonymous function in order to avoid labelled loop; I used return instead of `break`, since `break` would only break corresponding `select`.

anovakovic01 replied:
That is ok, but it seems to me that this should be separate (named) function in that case. IMHO it will improve code readability.

anovakovic01 commented on coap/nats/service.go:
```diff
@@ -0,0 +1,21 @@
+}
+
+// Subscription interface represents subscription to messaging system.
+type Subscription interface {
```
You can replace this with function type. Also, can you handle unsubscribe via channels?

mijicd replied:
Worth considering.

anovakovic01 commented on coap/server.go:
```diff
@@ -1,30 +1,107 @@
+		publisher, err := Authorize(&msg, respMsg, cid)
+		if err != nil {
+			respMsg.Code = gocoap.Unauthorized
+		} else {
```
Avoid `else` by using `break`.

anovakovic01 commented on cmd/coap/main.go:
```diff
@@ -6,44 +6,59 @@ import (
+	pub := nats.New(nc)
 
+	mgr := manager.NewClient(cfg.ManagerURL)
+	svc := coap.New(pub)
```
Add logging and metrics.

mijicd replied:
Make sure this remark by @anovakovic01 is implemented. See websocket adapter for recommended metrics.

anovakovic01 commented on coap/adapter.go:
```diff
@@ -1,61 +1,85 @@
+		svc.Unsubscribe(clientID)
+	}
+	svc.mu.Lock()
+	svc.Subs[clientID] = nats.Observer{
```
Maybe you should encapsulate set and delete methods for this map that will be thread safe (use lock).

anovakovic01 commented on coap/adapter.go:
```diff
@@ -1,61 +1,85 @@
+		return err
+	}
+	svc.mu.Lock()
+	delete(svc.Subs, id)
```
Same as for setting value.

anovakovic01 commented on coap/server.go:
```diff
@@ -1,30 +1,107 @@
-package adapter
+package coap
```
Maybe this file should be part of `api` package?

anovakovic01 commented on coap/service.go:
```diff
@@ -0,0 +1,28 @@
+	// service map of subscriptions under given ID.
+	Subscribe(string, string, chan mainflux.RawMessage) error
+	// Unsubscribe method is used to stop observing resource.
+	Unsubscribe(string) error
```
Can you avoid `Unsubscribe` method by using channel to unsubscribe?

dborovcanin replied:
Yes, but I didn't want to enforce that interface where it is not necessary. I already got `Subscription` from NATS client and I've got a map for CoAP purposes, so I didn't want to enforce that kind of unsubscription handling and to make one more goroutine if it isn't really necessary. Instead, I use what I've already got.

mijicd replied:
I have already suggested renaming this function. However, I am not sure whether it should be exported or not. Let's try handling this case via channels (message passing), and renaming it if needed. If that works, we can safely "unexport" it.

dborovcanin committed:
Add message id to notification masseges

Remove conflict in Gopkg.lock.
Change way observe option value is generated.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

drasko commented on coap/api/transport.go:
```diff
@@ -0,0 +1,206 @@
+	for i := 0; i < 3; i++ {
+		err = gocoap.Transmit(conn, addr, *msg)
+		if err != nil {
+			time.Sleep(5 * time.Millisecond)
```
Where does this 5ms comes from? Spec?

drasko commented on coap/api/transport.go:
```diff
@@ -0,0 +1,206 @@
+	}
+	msg.MessageID = *counter
+	msg.SetOption(gocoap.Observe, buff.Bytes()[:3])
+	for i := 0; i < 3; i++ {
```
Where does this 3 times comes from? Spec?

drasko commented on coap/server.go:
```diff
@@ -1,30 +1,114 @@
 
-	r.NotFoundHandler = coap.FuncHandler(notFoundHandler)
+// ListenAndServe binds to the given address and serve requests forever.
+func ListenAndServe(svc Service, mgr manager.ManagerClient, addr string, rh gocoap.Handler) error {
```
You are oveloading Dustin's Go server. This is OK for now - probably this way will be easier to bump up his version.

However, on the long-run, I think we'll need to rewrite this server because it is pretty unmaintained and has MIT license.

dborovcanin committed:
Remove message ID tracking in observing resources

Since only token matters, message ID is only copied from original message.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add lock when reading from the map of obsevers

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix way of generating observe option

Fix generating observe option to match protocol specification.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add timeout to confirmable message

In case of confirmable message, timeout is used in case client's
ACK or RST message is lost in transport.
Refactor code.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin reviewed (commented):
@drasko As I know, protocol spec is not explicit when it comes to a number of repeats before server gives up on notifying the client, nor about timeout between server's confirmable notification and corresponding acknowledgement or reject message from the client. There is no specific reason I took this approach, it looks just fine to me at the moment.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,203 @@
+			if msg.IsConfirmable() {
+				res.Code = gocoap.InternalServerError
+			}
+			return res
```
This return statement is overhead.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,203 @@
+
+func notify(svc coap.Service, id string, conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) error {
+	if err := sendMessage(conn, addr, msg); err != nil {
+		return svc.Unsubscribe(id)
```
Maybe return `sendMessage` error instead of returning `Unsubscribe` error.

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -0,0 +1,203 @@
+	now := time.Now().UnixNano() / timestamp
+	buff := new(bytes.Buffer)
+	err = binary.Write(buff, binary.LittleEndian, now)
+	if err != nil {
```
Can be part of `if` statement.

anovakovic01 commented on coap/server.go:
```diff
@@ -1,30 +1,129 @@
+		publisher, err := Authorize(&msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+		} else {
```
Avoid `else` by using `break`.

dborovcanin replied:
Using `break` is probably faster, but since this is typical `if-else` case, I'd prefer to keep it this way.

anovakovic01 commented on coap/server.go:
```diff
@@ -1,30 +1,129 @@
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			svc.RemoveTimeout(id)
+			err := svc.Unsubscribe(id)
+			if err != nil {
```
Can be part of `if` statement.

dborovcanin committed:
Remove notify method

Notify method replaced by explicit sendMessage and Unsubscribe method calls.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Move server.go to api package

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Change Authorize method to authorize

No need to export this method.
Refactored code.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

anovakovic01 commented on coap/api/server.go:
```diff
@@ -0,0 +1,126 @@
+		publisher, err := authorize(&msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+		} else {
```
From [Effective Go](https://golang.org/doc/effective_go.html#if): `In the Go libraries, you'll find that when an if statement doesn't flow into the next statement—that is, the body ends in break, continue, goto, or return—the unnecessary else is omitted.` We should follow this guide and be consistent.

mijicd replied:
I agree with this. Please omit `else`s wherever possible.

mijicd commented on coap/api/transport.go:
```diff
@@ -0,0 +1,192 @@
+	return r
+}
+
+func receive(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
```
Make a type alias for this return value.

mijicd commented on coap/api/transport.go:
```diff
@@ -0,0 +1,192 @@
+	}
+}
+
+func observe(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
```
Reuse it here.

mijicd commented on cmd/coap/main.go:
```diff
@@ -6,44 +6,59 @@ import (
 
-	logger, _ := zap.NewProduction()
-	defer logger.Sync() // flushes buffer, if any
+	logger := log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
```
Replace with mainflux logger and alight logging with other services.

mijicd commented on coap/README.md:
```diff
@@ -9,9 +9,11 @@ The service is configured using the environment variables presented in the
+|-----------------------|------------------------|-------------------------|
+| MF_COAP_ADAPTER_PORT  | adapter listening port | `5683`                  |
+| MF_NATS_URL           | NATS instance URL      | `nats://localhost:4222` |
+| FM_MANAGER_URL        | manager service URL    | `http://localhost:8180` |
```
Fix type, `MF` instead of `FM`.

mijicd commented on coap/README.md:
```diff
@@ -25,28 +27,16 @@ services:
-
-# set the environment variables and run the service
-COAP_ADAPTER_NATS_URL=[NATS instance URL] app
+make coap
```
Please ensure that README content is aligned with other README files.

dborovcanin committed:
Replace anonymous function with labeled for loop

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
+// New creates new CoAP adapter service struct.
+func New(pubsub nats.Service) Service {
+	return &adapterService{
+		pubsub: pubsub,
```
Please rename this `pubsub` to `broker` or `bus`. Apply the same change to websocket adapter, if present.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
-	ca.obsTransmit(m)
+func (svc *adapterService) RemoveTimeout(clientID string) {
+	svc.mu.Lock()
+	sub, ok := svc.subs[clientID]
```
This can be inlined:

```golang
if sub, ok := svc.subs[clientID]; ok {
    sub.Timeout <- false
}
```

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
+	if !ok {
+		return errEntryNotFound
 	}
+	timer := time.NewTimer(timeout)
```
Timer can be passed as parameter to this goroutine.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
+// AdapterService struct represents CoAP adapter service implementation.
+type adapterService struct {
+	pubsub nats.Service
+	subs   map[string]nats.Observer
```
It make sense  to "lift" this map as synchronized one, since all access patterns include locking/unlocking the accompanying mutex.

As the matter of fact, instead of building your own type, Go's `sync` package provides you the concurrent map.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
+		return err
+	}
+	svc.mu.Lock()
+	svc.subs[clientID] = nats.Observer{
```
Please align naming with ws adapter.

mijicd commented on Gopkg.lock:
```diff
@@ -324,6 +324,7 @@
   version = "v1.1.1"
 
 [[projects]]
+<<<<<<< HEAD
```
... :)

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
 
-const protocol string = "coap"
+const (
+	key     string = "key"
```
It seems there are no usages for these constants.

mijicd commented on coap/adapter.go:
```diff
@@ -1,61 +1,116 @@
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+func (svc *adapterService) Subscribe(chanID, clientID string, ch chan mainflux.RawMessage) error {
```
Technically, you're not unsubscribing but replacing the connection, as far as I understand (with resource cleanup). Since that's the case, I would consider making it explicit through function naming (and their implementation), and rename the `Unsubscribe` to `CreateOrReplaceConnection`.

mijicd commented on coap/api/server.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+func authKey(opt interface{}) (string, error) {
+	if opt == nil {
```
You can remove this check. It is also performed as part of the type assertion below.

mijicd commented on coap/api/server.go:
```diff
@@ -0,0 +1,126 @@
+		return "", errBadRequest
+	}
+	arr := strings.Split(val, "=")
+	if len(arr) != 2 || strings.ToLower(arr[0]) != "key" {
```
I suppose this was the place to use that unused constant defined above.

mijicd commented on coap/api/server.go:
```diff
@@ -0,0 +1,126 @@
+	}
+	switch msg.Type {
+	case gocoap.Reset:
+		cid := mux.Var(&msg, "id")
```
And I assume this is the place where the second constant should have been used.

mijicd commented on coap/api/transport.go:
```diff
@@ -0,0 +1,191 @@
+	errBadOption  = errors.New("bad option")
+	auth          manager.ManagerClient
+	maxPktLen     = 1500
+	network       = "udp"
```
Shouldn't these be defined as constants?

mijicd commented on coap/api/transport.go:
```diff
@@ -0,0 +1,191 @@
+	}
+	msg.SetOption(gocoap.Observe, buff.Bytes()[:3])
+	timeout := time.Duration(5)
+	// Try to transmit 3 times; each time duplicate timeout between attempts.
```
Is this specified in RFC? If so, then proceed. Otherwise, consider utilizing [exponential backoff](https://github.com/cenkalti/backoff).

mijicd commented on coap/nats/pubsub.go:
```diff
@@ -0,0 +1,64 @@
+
+type natsPublisher struct {
+	nc *broker.Conn
+	cb *gobreaker.CircuitBreaker
```
While it is perfectly suitable place for a circuit breaker, I would advise that we drop it for now, for the sake of making the code slightly simpler.

Please note that this might be changed in the future.

mijicd commented on coap/nats/pubsub.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (pubsub *natsPublisher) Subscribe(chanID string, ch chan mainflux.RawMessage) (Subscription, error) {
+	var sub *broker.Subscription
```
No need for declaring it in advance.

mijicd commented on coap/nats/service.go:
```diff
@@ -0,0 +1,22 @@
+// Service specifies NATS service API.
+type Service interface {
+	mainflux.MessagePublisher
+	// Subscribes to channel with specified id.
```
Follow the godoc recommendations.

mijicd reviewed (changes requested):
Do the proper rebase (resolve all conflicts). Prune the dependencies. Align with logging and naming used in other services.

dborovcanin committed:
Fix conflict in Gopkg.lock

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Replace Go-Kit logger with Mainflux logger

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix README.md typo

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove unused constants in adapter.go

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove lock-unlock blocks from adapter service

In order to get, put or remove values to/from subscribers map, functions wrapping those lock-unlock blocks are used.
Intentionally, use of defer is avoided in order to prevent performance drop.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix generating value of observe option

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix retransmission to match RFC spec

Retransmission fixed to use MaxRetransmit header when exists or default perams in case of
absence of MaxRetransmit header. Sending confirmable message timer set to 12 hours.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove circuit breaker from NATS broker

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove Zap logger from vendor

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Use channels to handle unsubscribe

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Update README and Makefile

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add timeout change when retransmitting confirmable message

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove unused max-age option from respose

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

drasko commented on coap/nats/broker.go:
```diff
@@ -0,0 +1,65 @@
+
+	go func() {
+		<-channel.Closed
+		for i := 0; i < 3; i++ {
```
I still do not understand where this number 3 comes from.

drasko replied:
Btw. did you want to use `maxFailedReqs` const here?

mijicd replied:
If it is specified by standard, which I assume, please define it as a constant, instead of a magic value.

dborovcanin committed:
Remove unnecessary for loop in case of unsubscribe error

Since error can ocure only in case of trying to unsubscribe from already unsubscribed
subscription or in case of closed connection. Either way, there are no need for multiple
unsubscription attempts.
Remove unused Subscription interface and unused constants in broker.go.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Don't copy maxRetransmit option from request message

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add timeouts for confirmable messages

Timeouts are explained in this section https://tools.ietf.org/html/rfc7252#page-28.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add logging and metrics middleware

Change timeout for sending confirmable message to 12H.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>
